tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -344,8 +377,18 @@ const DeviceCommand EnqueueReadBufferCommand::assemble_device_command(uint32_t d
     command.set_consumer_cb_size(consumer_cb_size);
     command.set_producer_cb_num_pages(producer_cb_num_pages);
     command.set_consumer_cb_num_pages(consumer_cb_num_pages);
-    command.set_num_pages(this->buffer.num_pages());
-    command.set_data_size(padded_page_size * this->buffer.num_pages());
+    command.set_num_pages(num_pages);
+    command.set_data_size(padded_page_size * num_pages);
+
+    if (is_sharded(this->buffer.buffer_layout())) {
+        uint32_t num_cores = this->buffer.num_cores();
+        auto core_bank_indices = this->buffer.core_bank_indices();
+        command.set_num_cores(num_cores);
```

Comment:
Given this is only used in the sharded path, probably calling this set_sharded_buffer_num_cores would be a more fitting name. And then in the kernels, you pull out sharded_buffer_num_cores from the device command.